### ISSUE TITLE
Cordova build iOS error: Cannot read property 'toLowerCase' of undefined with Xcode 9.4.1

### ISSUE THREAD

**Issue by lynzz** [apache/cordova-ios#427]
Ionic:

   ionic (Ionic CLI)  : 4.0.2 (/Users/lzz/.nvm/versions/node/v8.9.1/lib/node_modules/ionic)
   Ionic Framework    : ionic-angular 3.9.2
   @ionic/app-scripts : 3.1.8

Cordova:

   cordova (Cordova CLI) : 7.1.0
   Cordova Platforms     : android 6.3.0, ios 4.5.5

System:

   Android SDK Tools : 25.2.3
   ios-deploy        : 1.9.2
   ios-sim           : 6.1.2
   NodeJS            : v8.9.1 (/Users/lzz/.nvm/versions/node/v8.9.1/bin/node)
   npm               : 5.5.1
   OS                : macOS Mojave 10.14.1 Beta
   Xcode             : Xcode 9.4 Build version 9F1027a

Environment:

   ANDROID_HOME : /Users/lzz/Library/Android/sdk

Reinstall xcode 10, and such a mistake

Why?

**Comment by kashban** [comment]
I already made a Pull request to fix this:

https://github.com/apache/cordova-ios/pull/428

**Comment by lynzz** [comment]
@kashban Thanks, why would I use xcode 9.4 reported that such a mistake，for macOS 10.14.1 Beta ?

**Comment by kashban** [comment]
There was an update for the xcode build tools not long ago which was also issued for XCode 10 (not beta), breaking the same stuff. Perhaps it's not bound to Xcode 10.1 beta.

**Comment by csabbey** [comment]
I'm getting the below stack trace. My list-emulator-build-targets line 54 is:
`device.availability.toLowerCase().indexOf('unavailable') < 0`

It looks like the device availability property got renamed to isAvailable; here's one of the devices:

```
{ state: 'Shutdown',
  isAvailable: 'YES',
  name: 'iPad Pro (12.9-inch)',
  udid: '5A707375-7596-4375-9A7A-75690B681B46',
  availabilityError: '' }
```

... and here's the promised stack trace:

```
TypeError: Cannot read property 'toLowerCase' of undefined
    at /Users/craig.sabbey/code/ms-client-echidna/platforms/ios/cordova/lib/list-emulator-build-targets:54:45
    at Array.forEach (<anonymous>)
    at /Users/craig.sabbey/code/ms-client-echidna/platforms/ios/cordova/lib/list-emulator-build-targets:52:44
    at Array.reduce (<anonymous>)
    at /Users/craig.sabbey/code/ms-client-echidna/platforms/ios/cordova/lib/list-emulator-build-targets:50:57
    at Array.reduce (<anonymous>)
    at /Users/craig.sabbey/code/ms-client-echidna/platforms/ios/cordova/lib/list-emulator-build-targets:45:28
    at _fulfilled (/Users/craig.sabbey/code/ms-client-echidna/platforms/ios/cordova/node_modules/q/q.js:854:54)
    at /Users/craig.sabbey/code/ms-client-echidna/platforms/ios/cordova/node_modules/q/q.js:883:30
    at Promise.promise.promiseDispatch (/Users/craig.sabbey/code/ms-client-echidna/platforms/ios/cordova/node_modules/q/q.js:816:13)
```

**Comment by kashban** [comment]
That's exactly the issue my PR fixes.

**Comment by csabbey** [comment]
@kashban Yes, it looks like the same issue.

**Comment by waflessnet** [comment]
@csabbey , yes, i  change the "if" line  54 of platforms/ios/cordova/lib/list-emulator-build-targets  for: 
```javascript
if (device.name === deviceType.name.replace(/\-inch/g, ' inch') &&  device.isAvailable == 'YES') 
```
then, the function "availableDevices"  is:
```javascript
 var availableDevices = Object.keys(devices).reduce(function (availAcc, deviceCategory) {
                var availableDevicesInCategory = devices[deviceCategory];
                availableDevicesInCategory.forEach(function (device) {
                        //console.log(device);
                    if (device.name === deviceType.name.replace(/\-inch/g, ' inch') && device.isAvailable == 'YES') {
                            availAcc.push(device);
                        }
                });
                return availAcc;
            }, []);

```
It works!

**Comment by kashban** [comment]
@waflessnet Once my PR is merged and released with the next major version the change will be permanent. It even considers both json formats, so it will be backward compatible.

**Comment by newzhongdan** [comment]
in xcode > 10 dont work :(

**Comment by kennybky** [comment]
I had this same problem, but I found a solution. I don't know if it would work for others but I should probably share it in case.
I uninstalled  then reinstalled ios platform.
I ran npm update. 


Note: I am using cordova without ionic

Cordova: 8.1.2 (cordova-lib@8.1.1)
ios-deploy: 1.9.2
xcode: 9.4.1

**Comment by kashban** [comment]
Shouldn't happen anymore with XCode 10.1. Apple reintroduced the former availability property and added a new boolean one called isAvailable.

This accepted merge request takes care of both and will be in the next major release: https://github.com/apache/cordova-ios/pull/451

**Comment by Ramprasathselvam** [comment]
```
K-MacBook-Pro:projectname2.1.7(2.1.7.3) K$ cordova build ios
Cannot read property 'toLowerCase' of undefined
```
any solution for above issue.

**Comment by joeldhenry** [comment]
this started happening on xcode 10.2.1 and i have no solution to get it working again :(

**Comment by janpio** [comment]
Create a new issue with all the ncessary information please. Thanks.

**Comment by lucasgalli** [comment]
it's happening in Xcode 10.2.1 and i didn't find any solution.

**Comment by JillevdW** [comment]
It just happened to me with Xcode 10.2, running on the MacOS Catalina beta. I fixed it by slightly altering the answer @waflessnet gave. Replace the `availableDevices` function in `platforms/ios/cordova/lib/list-emulator-build-targets` with the following:

    var availableDevices = Object.keys(devices).reduce(function (availAcc, deviceCategory) {
        var availableDevicesInCategory = devices[deviceCategory];
        availableDevicesInCategory.forEach(function (device) {
            if (device && device.name === deviceType.name.replace(/\-inch/g, ' inch') && device.isAvailable == true) {
                availAcc.push(device);
            }
        });
        return availAcc;
    }, []);

**Comment by lucasgalli** [comment]
I don't know why but I had already tried and just worked it after that I reinstall the Xcode, anyway now it's working, thanks a lot @JillevdW and @waflessnet.

**Comment by Matttaylor8910** [comment]
Why is this closed? This still happens with the latest cordova release installed

**Comment by brodybits** [comment]
Please raise a new issue with the necessary information, as requested by @janpio. We do not actively support discussions in closed issues. It is fine for the new issue to reference this issue number.

And please keep in mind that this project is maintained by a bunch of overworked volunteers.

**Comment by derman10** [comment]
> It just happened to me with Xcode 10.2, running on the MacOS Catalina beta. I fixed it by slightly altering the answer @waflessnet gave. Replace the `availableDevices` function in `platforms/ios/cordova/lib/list-emulator-build-targets` with the following:
> 
> ```
> var availableDevices = Object.keys(devices).reduce(function (availAcc, deviceCategory) {
>     var availableDevicesInCategory = devices[deviceCategory];
>     availableDevicesInCategory.forEach(function (device) {
>         if (device && device.name === deviceType.name.replace(/\-inch/g, ' inch') && device.isAvailable == true) {
>             availAcc.push(device);
>         }
>     });
>     return availAcc;
> }, []);
> ```

this solved my problem! 
Thanks!!

**Comment by ahmadalfy** [comment]
I have cordova `9.0.0 (cordova-lib@9.0.1)` and xCode `Version 11.0 (11A420a)`. Still facing this issue

**Comment by renatoflorencia** [comment]
I am experiencing the same problem after updating xcode.
My current project has the following versions:

cordova: 8.0.0
ios platform: ^ 4.5.5
xcode: 11.0
ios version: Mojave 10.14

**Comment by brodybits** [comment]
Guys please report this kind of thing in a *new* issue with a [reproducible example ([1])](https://stackoverflow.com/help/minimal-reproducible-example) as we do not support discussions in closed issues.

[1] https://stackoverflow.com/help/minimal-reproducible-example

**Comment by osmanc777** [comment]
![Captura de Pantalla 2019-09-25 a la(s) 16 19 34](https://user-images.githubusercontent.com/23218509/65714577-a1ee3e80-e058-11e9-9f5d-8a9475ae0218.png)

Hello @ahmadalfy @renatoflorencia and @brodybits , this error arose after the last update of Xcode since before if it worked normally. The Xcode version I try is the following * 11.0 (11A420a) *

I have tried to execute several commands and modify the build files of the project (ios) but I have not obtained any positive results

List of commands executed:
`ionic cordova build ios --prod`
`ionic cordova build ios`
`ionic cordova build ios - - buildFlag = "UseModernSystem = 0`
`ionic cordova build --prod --release ios`
`ionic cordova build ios --release --prod --device --buildFlag = "- UseModernBuildSystem = 0" --packageType = "app-store`
`ionic cordova build ios --release --device --buildFlag = "- UseModernBuildSystem = 0" --packageType = "app-store"`

and none of these worked for a successful compilation.

**Comment by brodybits** [comment]
No more discussion on this issue please. Please raise a *new* issue with the reproducible example.

At this point the maintainers to not have time to actively support use with Ionic. Please consider using Capacitor for Ionic, since Capacitor was designed to work more smoothly with Ionic.

I am now locking this issue as resolved.
